pkg/genlib/generator_interface.go

```diff
@@ -0,0 +1,289 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package genlib
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/Pallinder/go-randomdata"
+	"github.com/elastic/elastic-integration-corpus-generator-tool/pkg/genlib/config"
+	"github.com/elastic/elastic-integration-corpus-generator-tool/pkg/genlib/fields"
+	"github.com/lithammer/shortuuid/v3"
+	"math"
+	"math/rand"
+	"regexp"
+	"strings"
+	"sync"
+)
+
+type (
+	Fields      = fields.Fields
+	Field       = fields.Field
+	Config      = config.Config
+	ConfigField = config.ConfigField
+)
+
+const (
+	FieldTypeBool            = "boolean"
+	FieldTypeKeyword         = "keyword"
+	FieldTypeConstantKeyword = "constant_keyword"
+	FieldTypeDate            = "date"
+	FieldTypeIP              = "ip"
+	FieldTypeDouble          = "double"
+	FieldTypeFloat           = "float"
+	FieldTypeHalfFloat       = "half_float"
+	FieldTypeScaledFloat     = "scaled_float"
+	FieldTypeInteger         = "integer"
+	FieldTypeLong            = "long"
+	FieldTypeUnsignedLong    = "unsigned_long"
+	FieldTypeObject          = "object"
+	FieldTypeNested          = "nested"
+	FieldTypeFlattened       = "flattened"
+	FieldTypeGeoPoint        = "geo_point"
+
+	FieldTypeTimeRange  = 3600 // seconds
+	FieldTypeTimeLayout = "2006-01-02T15:04:05.999999Z07:00"
+)
+
+var (
+	replacer     = strings.NewReplacer(".*", "")
+	keywordRegex = regexp.MustCompile("(\\.|-|_|\\s){1,1}")
+)
+
+// Typedef of the internal emit function
+type emitF func(state *GenState, dupes map[string]struct{}, buf *bytes.Buffer) error
+
+type Generator interface {
```

Comment:
`Generator` is now an interface: we have to implementation `GenerateJson` (previously `Generate` non-interface) and `GeneratorWithTemplate`

btw, the two concrete implementations probably can be made private